Core/fft_mt_r2iq.cpp

```diff
@@ -180,38 +178,11 @@ void fft_mt_r2iq::Init(float gain, int16_t **buffers, float** obuffers)
 			filterHw[d] = (fftwf_complex*)fftwf_malloc(sizeof(fftwf_complex)*halfFft);     // halfFft
 		}
 
-		for (int t = 0; t < halfFft; t++)
-		{
-			pfilterht[t][0] = 0.0f;
-			pfilterht[t][1] = 0.0f;
-		}
-
 		filterplan_t2f_c2c = fftwf_plan_dft_1d(halfFft, pfilterht, filterHw[0], FFTW_FORWARD, FFTW_MEASURE);
 		float *pht = new float[halfFft / 4 + 1];
 		for (int d = 0; d < NDECIDX; d++)
 		{
-			switch (d)
-			{
-			case 5:
-				KaiserWindow(halfFft / 4 + 1, 80.0f, 0.7f/64.0f, 1.0f/64.0f, pht);
-				break;
-			case 4:
-				KaiserWindow(halfFft / 4 + 1, 90.0f, 1.6f/64.0f, 2.0f/64.0f, pht);
-				break;
-			case 3:
-				KaiserWindow(halfFft / 4 + 1, 100.0f, 3.6f/64.0f, 4.0f/64.0f, pht);
-				break;
-			case 2:
-				KaiserWindow(halfFft / 4 + 1, 110.0f, 7.6f/64.0f, 8.0f/64.0f, pht);
-				break;
-			case 1:
-				KaiserWindow(halfFft / 4 + 1, 120.0f, 15.7f/64.0f, 16.0f/64.0f, pht);
-				break;
-			case 0:
-			default:
-				KaiserWindow(halfFft / 4 + 1, 120.0f, 30.5f/64.0f, 32.0f/64.0f, pht);
-				break;
-			}
+			KaiserWindow(halfFft / 4 + 1, 120.0f, 61.0f/mratio[d]/128.0f, 64.0f/mratio[d]/128.0f, pht);
```

Comment:
i'm unsure about the side effects .. and if the KaiserWindow() surely can't provide/the 120 dB attenuation.
i'll test this!

Reply:
okay. the code of KaiserWindow is copied from CuteSDR. just for your information.

Reply:
usually, you tell the window design function to generate a filter fulfilling criteria as passband and stopband-attenuation. then the design algorithm tries to find the minimum number of taps meeting these.

other way is, to give passband and number of taps .. and it creates "best-possible" coefficients for the given number of taps.

however, giving stopband-attenuation **and** the number of taps .. i assume it falls back to 2nd case.
this is why we need to use dynamic filter size - depending on the decimation.

Reply:
i am not sure how to handle dynamic filter size when applying filter in freq domain. I need to read math book again.

Reply:
we use a Ht length of FFTN/4+1 that possibly is the max to easily apply overlap/scrap.
We have to increase FFTN,   With FFT order 4096 ht is 1025 and it hard at our rate to have sharp filter with 1 MHz or lower cutoff.  If we cannot increase FFT order we could use two decimation stages. First to the double of final rate with a less demanding filter and then a second one in to the final goal maybe this second one could be done in time ? ?

Reply:
i have played around with the KaiserWindow() do check it's estimated number of required coefficients:
```
***************************************************************************
Filter tap estimation
decimation  0: KaiserWindow(Astop = 120.0 dB, Fpass = 61.000000 @ 128 Hz,Fstop = 64.000000 @ 128 Hz) => 333 taps
decimation  1: KaiserWindow(Astop = 120.0 dB, Fpass = 30.500000 @ 128 Hz,Fstop = 32.000000 @ 128 Hz) => 666 taps
decimation  2: KaiserWindow(Astop = 120.0 dB, Fpass = 15.250000 @ 128 Hz,Fstop = 16.000000 @ 128 Hz) => 1332 taps
decimation  3: KaiserWindow(Astop = 120.0 dB, Fpass = 7.625000 @ 128 Hz,Fstop = 8.000000 @ 128 Hz) => 2663 taps
decimation  4: KaiserWindow(Astop = 120.0 dB, Fpass = 3.812500 @ 128 Hz,Fstop = 4.000000 @ 128 Hz) => 5326 taps
decimation  5: KaiserWindow(Astop = 120.0 dB, Fpass = 1.906250 @ 128 Hz,Fstop = 2.000000 @ 128 Hz) => 10652 taps
decimation  6: KaiserWindow(Astop = 120.0 dB, Fpass = 0.953125 @ 128 Hz,Fstop = 1.000000 @ 128 Hz) => 21303 taps
decimation  7: KaiserWindow(Astop = 120.0 dB, Fpass = 0.476563 @ 128 Hz,Fstop = 0.500000 @ 128 Hz) => 42605 taps
decimation  8: KaiserWindow(Astop = 120.0 dB, Fpass = 0.238281 @ 128 Hz,Fstop = 0.250000 @ 128 Hz) => 85209 taps
decimation  9: KaiserWindow(Astop = 120.0 dB, Fpass = 0.119141 @ 128 Hz,Fstop = 0.125000 @ 128 Hz) => 170417 taps
decimation 10: KaiserWindow(Astop = 120.0 dB, Fpass = 0.059570 @ 128 Hz,Fstop = 0.062500 @ 128 Hz) => 340833 taps
decimation 11: KaiserWindow(Astop = 120.0 dB, Fpass = 0.029785 @ 128 Hz,Fstop = 0.031250 @ 128 Hz) => 681665 taps
decimation 12: KaiserWindow(Astop = 120.0 dB, Fpass = 0.014893 @ 128 Hz,Fstop = 0.015625 @ 128 Hz) => 1363329 taps
decimation 13: KaiserWindow(Astop = 120.0 dB, Fpass = 0.007446 @ 128 Hz,Fstop = 0.007813 @ 128 Hz) => 2726658 taps
decimation 14: KaiserWindow(Astop = 120.0 dB, Fpass = 0.003723 @ 128 Hz,Fstop = 0.003906 @ 128 Hz) => 5453316 taps
decimation 15: KaiserWindow(Astop = 120.0 dB, Fpass = 0.001862 @ 128 Hz,Fstop = 0.001953 @ 128 Hz) => 10906631 taps
***************************************************************************
```
you see, that with _current_ specification, the number of taps would explode.

in addition, NDECIDX=16, is way too much. it would decimate downto ~ 2 kHz from 128 MHz. NDECIDX=10 should be sufficient.

then, i played a bit with the filter specifications .. and optimized the debug output:
```
***************************************************************************
Filter tap estimation, Astop = 120.0 dB, relPass = 0.90, relStop = 1.05
decimation  0: KaiserWindow(Astop = 120.0 dB, Fpass = 57.600,Fstop = 67.200, Bw 64.000 @ 128.000000 ) => 105 taps
decimation  1: KaiserWindow(Astop = 120.0 dB, Fpass = 28.800,Fstop = 33.600, Bw 32.000 @ 128.000000 ) => 209 taps
decimation  2: KaiserWindow(Astop = 120.0 dB, Fpass = 14.400,Fstop = 16.800, Bw 16.000 @ 128.000000 ) => 417 taps
decimation  3: KaiserWindow(Astop = 120.0 dB, Fpass = 7.200,Fstop = 8.400, Bw 8.000 @ 128.000000 ) => 833 taps
decimation  4: KaiserWindow(Astop = 120.0 dB, Fpass = 3.600,Fstop = 4.200, Bw 4.000 @ 128.000000 ) => 1665 taps  <25% of 8192
decimation  5: KaiserWindow(Astop = 120.0 dB, Fpass = 1.800,Fstop = 2.100, Bw 2.000 @ 128.000000 ) => 3329 taps
decimation  6: KaiserWindow(Astop = 120.0 dB, Fpass = 0.900,Fstop = 1.050, Bw 1.000 @ 128.000000 ) => 6657 taps
decimation  7: KaiserWindow(Astop = 120.0 dB, Fpass = 0.450,Fstop = 0.525, Bw 0.500 @ 128.000000 ) => 13314 taps
decimation  8: KaiserWindow(Astop = 120.0 dB, Fpass = 0.225,Fstop = 0.262, Bw 0.250 @ 128.000000 ) => 26628 taps
decimation  9: KaiserWindow(Astop = 120.0 dB, Fpass = 0.112,Fstop = 0.131, Bw 0.125 @ 128.000000 ) => 53256 taps
***************************************************************************

***************************************************************************
Filter tap estimation, Astop = 110.0 dB, relPass = 0.90, relStop = 1.05
decimation  0: KaiserWindow(Astop = 110.0 dB, Fpass = 57.600,Fstop = 67.200, Bw 64.000 @ 128.000000 ) => 95 taps
decimation  1: KaiserWindow(Astop = 110.0 dB, Fpass = 28.800,Fstop = 33.600, Bw 32.000 @ 128.000000 ) => 190 taps
decimation  2: KaiserWindow(Astop = 110.0 dB, Fpass = 14.400,Fstop = 16.800, Bw 16.000 @ 128.000000 ) => 379 taps
decimation  3: KaiserWindow(Astop = 110.0 dB, Fpass = 7.200,Fstop = 8.400, Bw 8.000 @ 128.000000 ) => 758 taps
decimation  4: KaiserWindow(Astop = 110.0 dB, Fpass = 3.600,Fstop = 4.200, Bw 4.000 @ 128.000000 ) => 1516 taps  <25% of 8192
decimation  5: KaiserWindow(Astop = 110.0 dB, Fpass = 1.800,Fstop = 2.100, Bw 2.000 @ 128.000000 ) => 3032 taps
decimation  6: KaiserWindow(Astop = 110.0 dB, Fpass = 0.900,Fstop = 1.050, Bw 1.000 @ 128.000000 ) => 6063 taps
decimation  7: KaiserWindow(Astop = 110.0 dB, Fpass = 0.450,Fstop = 0.525, Bw 0.500 @ 128.000000 ) => 12126 taps
decimation  8: KaiserWindow(Astop = 110.0 dB, Fpass = 0.225,Fstop = 0.262, Bw 0.250 @ 128.000000 ) => 24251 taps
decimation  9: KaiserWindow(Astop = 110.0 dB, Fpass = 0.112,Fstop = 0.131, Bw 0.125 @ 128.000000 ) => 48501 taps
***************************************************************************

***************************************************************************
Filter tap estimation, Astop = 120.0 dB, relPass = 0.85, relStop = 1.10
decimation  0: KaiserWindow(Astop = 120.0 dB, Fpass = 54.400,Fstop = 70.400, Bw 64.000 @ 128.000000 ) => 63 taps
decimation  1: KaiserWindow(Astop = 120.0 dB, Fpass = 27.200,Fstop = 35.200, Bw 32.000 @ 128.000000 ) => 125 taps
decimation  2: KaiserWindow(Astop = 120.0 dB, Fpass = 13.600,Fstop = 17.600, Bw 16.000 @ 128.000000 ) => 250 taps
decimation  3: KaiserWindow(Astop = 120.0 dB, Fpass = 6.800,Fstop = 8.800, Bw 8.000 @ 128.000000 ) => 500 taps
decimation  4: KaiserWindow(Astop = 120.0 dB, Fpass = 3.400,Fstop = 4.400, Bw 4.000 @ 128.000000 ) => 999 taps
decimation  5: KaiserWindow(Astop = 120.0 dB, Fpass = 1.700,Fstop = 2.200, Bw 2.000 @ 128.000000 ) => 1998 taps  <25% of 8192
decimation  6: KaiserWindow(Astop = 120.0 dB, Fpass = 0.850,Fstop = 1.100, Bw 1.000 @ 128.000000 ) => 3995 taps
decimation  7: KaiserWindow(Astop = 120.0 dB, Fpass = 0.425,Fstop = 0.550, Bw 0.500 @ 128.000000 ) => 7989 taps
decimation  8: KaiserWindow(Astop = 120.0 dB, Fpass = 0.213,Fstop = 0.275, Bw 0.250 @ 128.000000 ) => 15977 taps
decimation  9: KaiserWindow(Astop = 120.0 dB, Fpass = 0.106,Fstop = 0.138, Bw 0.125 @ 128.000000 ) => 31954 taps
***************************************************************************

***************************************************************************
Filter tap estimation, Astop = 120.0 dB, relPass = 0.80, relStop = 1.15
decimation  0: KaiserWindow(Astop = 120.0 dB, Fpass = 51.200,Fstop = 73.600, Bw 64.000 @ 128.000000 ) => 45 taps
decimation  1: KaiserWindow(Astop = 120.0 dB, Fpass = 25.600,Fstop = 36.800, Bw 32.000 @ 128.000000 ) => 90 taps
decimation  2: KaiserWindow(Astop = 120.0 dB, Fpass = 12.800,Fstop = 18.400, Bw 16.000 @ 128.000000 ) => 179 taps
decimation  3: KaiserWindow(Astop = 120.0 dB, Fpass = 6.400,Fstop = 9.200, Bw 8.000 @ 128.000000 ) => 357 taps
decimation  4: KaiserWindow(Astop = 120.0 dB, Fpass = 3.200,Fstop = 4.600, Bw 4.000 @ 128.000000 ) => 714 taps
decimation  5: KaiserWindow(Astop = 120.0 dB, Fpass = 1.600,Fstop = 2.300, Bw 2.000 @ 128.000000 ) => 1427 taps  <25% of 8192
decimation  6: KaiserWindow(Astop = 120.0 dB, Fpass = 0.800,Fstop = 1.150, Bw 1.000 @ 128.000000 ) => 2853 taps
decimation  7: KaiserWindow(Astop = 120.0 dB, Fpass = 0.400,Fstop = 0.575, Bw 0.500 @ 128.000000 ) => 5706 taps
decimation  8: KaiserWindow(Astop = 120.0 dB, Fpass = 0.200,Fstop = 0.287, Bw 0.250 @ 128.000000 ) => 11412 taps
decimation  9: KaiserWindow(Astop = 120.0 dB, Fpass = 0.100,Fstop = 0.144, Bw 0.125 @ 128.000000 ) => 22824 taps
***************************************************************************

***************************************************************************
Filter tap estimation, Astop = 120.0 dB, relPass = 0.75, relStop = 1.20
decimation  0: KaiserWindow(Astop = 120.0 dB, Fpass = 48.000,Fstop = 76.800, Bw 64.000 @ 128.000000 ) => 35 taps
decimation  1: KaiserWindow(Astop = 120.0 dB, Fpass = 24.000,Fstop = 38.400, Bw 32.000 @ 128.000000 ) => 70 taps
decimation  2: KaiserWindow(Astop = 120.0 dB, Fpass = 12.000,Fstop = 19.200, Bw 16.000 @ 128.000000 ) => 139 taps
decimation  3: KaiserWindow(Astop = 120.0 dB, Fpass = 6.000,Fstop = 9.600, Bw 8.000 @ 128.000000 ) => 278 taps
decimation  4: KaiserWindow(Astop = 120.0 dB, Fpass = 3.000,Fstop = 4.800, Bw 4.000 @ 128.000000 ) => 555 taps
decimation  5: KaiserWindow(Astop = 120.0 dB, Fpass = 1.500,Fstop = 2.400, Bw 2.000 @ 128.000000 ) => 1110 taps  <25% of 8192
decimation  6: KaiserWindow(Astop = 120.0 dB, Fpass = 0.750,Fstop = 1.200, Bw 1.000 @ 128.000000 ) => 2219 taps
decimation  7: KaiserWindow(Astop = 120.0 dB, Fpass = 0.375,Fstop = 0.600, Bw 0.500 @ 128.000000 ) => 4438 taps
decimation  8: KaiserWindow(Astop = 120.0 dB, Fpass = 0.188,Fstop = 0.300, Bw 0.250 @ 128.000000 ) => 8876 taps
decimation  9: KaiserWindow(Astop = 120.0 dB, Fpass = 0.094,Fstop = 0.150, Bw 0.125 @ 128.000000 ) => 17752 taps
***************************************************************************
```
using FFTN/2+1 would also be _easily_ possible, but i dislike this compromise.
in the end, we need a tradeoff .. i would suggest `Filter tap estimation, Astop = 120.0 dB, relPass = 0.85, relStop = 1.10` which will be good for decimations down to 2 MHz.
(i might be off by one for the decimation and the corresponding Fpass/Fstop!)